tests/std/tests/P0896R4_counted_iterator_death/test.cpp

```diff
@@ -0,0 +1,248 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#define _CONTAINER_DEBUG_LEVEL 1
+
+#include <cassert>
+#include <concepts>
+#include <iterator>
+#include <vector>
+
+#include <test_death.hpp>
+using namespace std;
+
+int globalArray[5]{10, 20, 30, 40, 50};
+int otherArray[5]{10, 20, 30, 40, 50};
+vector<int> checkedArray{10, 20, 30, 40, 50};
+using vit = vector<int>::iterator;
+
+struct simple_input_iter {
+    using value_type      = int;
+    using difference_type = ptrdiff_t;
+
+    value_type operator*() const {
+        return *_ptr;
+    };
+    simple_input_iter& operator++() {
+        ++_ptr;
+        return *this;
+    };
+    simple_input_iter operator++(int) {
+        simple_input_iter _tmp = *this;
+        ++_ptr;
+        return _tmp;
+    };
+
+    bool operator==(simple_input_iter const&) const = default;
+
+    int* _ptr = nullptr;
+};
+
+void test_case_construction_negative_length() {
+    counted_iterator<int*> cit{globalArray, -1}; // counted_iterator requires positive length n
+    (void) cit;
+}
+
+void test_case_operator_dereference_value_initialized_iterator() {
+    counted_iterator<int*> cit{}; // note: for IDL to work correctly, default-init and value-init are equivalent
+    (void) (*cit); // cannot dereference value-initialized counted_iterator
+}
+
+void test_case_operator_dereference_end_iterator() {
+    counted_iterator<int*> cit{globalArray, 0};
+    (void) (*cit); // cannot dereference end counted_iterator (per proposed resolution of LWG-3472)
+}
+
+void test_case_operator_preincrement_value_initialized_iterator() {
+    counted_iterator<int*> cit{};
+    ++cit; // cannot increment value-initialized counted_iterator
+}
+
+void test_case_operator_preincrement_value_initialized_input_iterator() {
+    counted_iterator<simple_input_iter> cit{};
+    ++cit; // cannot increment value-initialized counted_iterator
+}
+
+void test_case_operator_preincrement_after_end() {
+    counted_iterator<int*> cit{globalArray, 0};
+    ++cit; // cannot increment counted_iterator past end
+}
+
+void test_case_operator_preincrement_after_end_input_iterator() {
+    counted_iterator<simple_input_iter> cit{simple_input_iter{globalArray}, 0};
+    ++cit; // cannot increment counted_iterator past end
+}
+
+void test_case_operator_increment_value_initialized_iterator() {

```

Comment:
Naming style: I think that the functions named plain "increment" and "decrement" should say "postincrement" and "postdecrement". It seems confusing to imply that post is "normal" when it's actually the weird one. (It's fine for the comments to say just "increment" for both pre and post.)

Reply:
Done